File: sample-projects/kotlin/core/src/main/kotlin/com/example/core/Result.kt
package com.example.core

sealed class Result<out T> {
    data class Success<out T>(val value: T) : Result<T>()
    data class Failure(val error: String) : Result<Nothing>()
}

fun <T> Result<T>.getOrNull(): T? = when (this) {
    is Result.Success -> value
    is Result.Failure -> null
}

fun <T> Result<T>.getOrThrow(): T = when (this) {
    is Result.Success -> value
    is Result.Failure -> throw IllegalStateException(error)
}

File: sample-projects/ruby/lib/core/result.rb
# typed: strict
# frozen_string_literal: true

module Core
  class Result
    extend T::Sig
    extend T::Helpers

    sealed!

    sig { returns(T::Boolean) }
    def success?
      is_a?(Success)
    end

    sig { returns(T::Boolean) }
    def failure?
      is_a?(Failure)
    end
  end

  class Success < Result
    extend T::Sig

    sig { returns(NilClass) }
    def value
      nil
    end
  end

  class Failure < Result
    extend T::Sig

    sig { params(error: String).void }
    def initialize(error)
      super()
      @error = T.let(error, String)
    end

    sig { returns(String) }
    def error
      @error
    end
  end
end
